Assets/Newtonsoft.Json.UnityConverters.Tests/Scripting/ScriptableObjectTests.cs

```diff
@@ -0,0 +1,145 @@
+        {
+            // Arrange
+            var serializer = JsonSerializer.Create(UnityConverterInitializer.DefaultUnityConvertersSettings);
+            serializer.TypeNameHandling = typeNameHandling;
+
+            MockScriptableObject instance = ScriptableObject.CreateInstance<MockScriptableObject>();
+            var obj = new {
+                scriptableObject = (ScriptableObject)instance,
+            };
+
+            // Act
+            string result = Serialize(obj, serializer);
+
+            var jObject = JObject.Parse(result);
+            JToken scriptableObjectToken = jObject["scriptableObject"];
+
+            return scriptableObjectToken.Value<string>("$type");
+        }
+
+        [Test]
+        [TestCase(TypeNameHandling.All)]
+        [TestCase(TypeNameHandling.Objects)]
+        [TestCase(TypeNameHandling.Arrays)]
+        [TestCase(TypeNameHandling.Auto)]
+        public void DeserializeGivenTypeFromProperty(TypeNameHandling typeNameHandling)
+        {
+            ScriptableObject result = DeserializeWithTypeNameHandling(typeNameHandling);
+
+            // Assert
+            Assert.IsInstanceOf<MockScriptableObject>(result);
+        }
+
+        [Test]
+        [TestCase(TypeNameHandling.None)]
+        public void DeserializesGivenTypeFromSignature(TypeNameHandling typeNameHandling)
+        {
+            ScriptableObject result = DeserializeWithTypeNameHandling(typeNameHandling);
+
+            // Assert
+            Assert.IsNotInstanceOf<MockScriptableObject>(result);
+        }
+
+        private static ScriptableObject DeserializeWithTypeNameHandling(TypeNameHandling typeNameHandling)
+        {
+            // Arrange
+            var serializer = JsonSerializer.Create(UnityConverterInitializer.DefaultUnityConvertersSettings);
+            serializer.TypeNameHandling = typeNameHandling;
+
+            string input = $@"{{""$type"":""{GetTypeNameWithAssembly(typeof(MockScriptableObject))}""}}";
+
+            // Act
+            ScriptableObject result = Deserialize<ScriptableObject>(input, serializer);
+            return result;
+        }
+
+        protected override bool AreEqual([AllowNull] MockScriptableObject a, [AllowNull] MockScriptableObject b)
+        {
+            if (a is null && b is null)
+            {
+                return true;
+            }
+
+            if (a is null || b is null)
+            {
+                return false;
+            }
+
+            return a.name == b.name
+                && a.hideFlags == b.hideFlags;
+        }
+
+        private static MockScriptableObject CreateMockInstance(string name, HideFlags hideFlags)
+        {
+            MockScriptableObject instance = ScriptableObject.CreateInstance<MockScriptableObject>();
+
+            instance.name = name;
+            instance.hideFlags = hideFlags;
+
+            return instance;
+        }
+
+        private struct ExpectedSignature
```

Comment:
Codacy found an issue: [Implement 'IEquatable<T>' in value type 'ExpectedSignature'.](https://app.codacy.com/manual/jilleJr/Newtonsoft.Json-for-Unity.Converters/pullRequest?prid=5855822)